File: fabric-audit-agent/core/sla.js
import { firstSeenMap } from './accountability.js';

const SLA_DAYS = { Critical: 1, Warning: 7, Info: 30 };
const DAY_MS = 86_400_000;

/**
 * Annotate findings with SLA age vs target (from first-seen in history). Pure.
 * @param {object[]} findings  active findings (with .key, .score)
 * @param {Array<object>} history  chronological run records
 * @param {number} nowMs  current time in ms (0 disables)
 * @param {Record<string,number>} [slaDays]
 * @returns {object[]}
 */
export function assessSla(findings = [], history = [], nowMs = 0, slaDays = SLA_DAYS) {
  const firstSeen = firstSeenMap(history);
  return findings.map(f => {
    const fs = firstSeen[f.key];
    const target = slaDays[f.score?.level];
    if (!fs || target == null || nowMs <= 0) return f;
    const firstSeenMs = Date.parse(fs);
    if (!Number.isFinite(firstSeenMs)) return f;
    const ageDays = Math.floor((nowMs - firstSeenMs) / DAY_MS);
    return { ...f, sla: { ageDays, targetDays: target, breached: ageDays > target } };
  });
}

/** Summarize SLA breaches. Pure. */
export function summarizeSla(findings = []) {
  const breached = findings.filter(f => f.sla?.breached);
  return {
    breachedCount: breached.length,
    items: breached.map(f => ({ key: f.key, level: f.score?.level, ageDays: f.sla.ageDays, targetDays: f.sla.targetDays })),
  };
}

File: fabric-audit-agent-py/fabric_audit_agent/sla.py
"""SLA age vs target per finding. Port of ``core/sla.js``. Pure (time injected)."""
import math
from datetime import datetime, timezone
from .accountability import first_seen_map

_SLA_DAYS = {"Critical": 1, "Warning": 7, "Info": 30}
_DAY_MS = 86_400_000


def _parse_ms(s):
    if not isinstance(s, str):
        return None
    try:
        dt = datetime.fromisoformat(s.replace("Z", "+00:00"))
    except ValueError:
        return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.timestamp() * 1000


def assess_sla(findings=None, history=None, now_ms=0, sla_days=None):
    findings = findings or []
    history = history or []
    sla_days = sla_days if sla_days is not None else _SLA_DAYS
    first_seen = first_seen_map(history)
    out = []
    for f in findings:
        fs = first_seen.get(f.get("key"))
        target = sla_days.get((f.get("score") or {}).get("level"))
        if not fs or target is None or now_ms <= 0:
            out.append(f)
            continue
        first_seen_ms = _parse_ms(fs)
        if first_seen_ms is None:
            out.append(f)
            continue
        age_days = math.floor((now_ms - first_seen_ms) / _DAY_MS)
        out.append({**f, "sla": {"ageDays": age_days, "targetDays": target, "breached": age_days > target}})
    return out


def summarize_sla(findings=None):
    findings = findings or []
    breached = [f for f in findings if (f.get("sla") or {}).get("breached")]
    return {
        "breachedCount": len(breached),
        "items": [{"key": f.get("key"), "level": (f.get("score") or {}).get("level"), "ageDays": f["sla"]["ageDays"], "targetDays": f["sla"]["targetDays"]} for f in breached],
    }
